lib/util/parse.ts

```diff
@@ -44,87 +40,83 @@ export function parseTransports(transports: any[]) {
 
 async function parseStorage(storage: any) {
   const trie = new Trie()
-  const promises = []
-  // eslint-disable-next-line prefer-const
-  for (let [address, value] of Object.entries(storage)) {
-    value = util.rlp.encode(util.unpadBuffer(toBuffer(value as string)))
-    promises.push(
-      new Promise((resolve, reject) => {
-        trie.put(toBuffer(address), value, (err: Error) => {
-          if (err) return reject(err)
-          resolve()
-        })
-      })
-    )
+  for (const [address, value] of Object.entries(storage)) {
+    const key = Buffer.from(address, 'hex')
+    const val = rlp.encode(unpadBuffer(Buffer.from(value as string, 'hex')))
+    await trie.put(key, val)
   }
-  await Promise.all(promises)
   return trie
 }
 
 async function parseGethState(alloc: any) {
   const trie = new Trie()
-  const promises = []
   for (const [key, value] of Object.entries(alloc)) {
-    const address = toBuffer(key)
-    const account = new util.Account()
-    if ((value as any).balance) {
-      // TODO: convert to buffer w/ util.toBuffer()?
-      // @ts-ignore: account.balance is type Buffer, not BN
-      account.balance = new util.BN((value as any).balance.slice(2), 16)
+    const address = isHexPrefixed(key) ? toBuffer(key) : Buffer.from(key, 'hex')
+    const { balance, code, storage } = value as any
+    const account = new Account()
+    if (balance) {
+      // note: balance is a Buffer
+      account.balance = new BN(toBuffer(balance))
     }
-    if ((value as any).code) {
-      account.codeHash = util.keccak(util.toBuffer((value as any).code))
+    if (code) {
+      account.codeHash = keccak(toBuffer(code))
     }
-    if ((value as any).storage) {
-      account.stateRoot = (await parseStorage((value as any).storage)).root
+    if (storage) {
+      const storageTrie = await parseStorage(storage)
+      account.stateRoot = storageTrie.root
     }
-    promises.push(
-      new Promise((resolve, reject) => {
-        trie.put(address, account.serialize(), (err: Error) => {
-          if (err) return reject(err)
-          resolve()
-        })
-      })
-    )
+    await trie.put(address, account.serialize())
   }
-  await Promise.all(promises)
   return trie
 }
 
 async function parseGethHeader(json: any) {
-  return BlockHeader.fromHeaderData(
-    {
-      gasLimit: new util.BN(util.stripHexPrefix(json.gasLimit), 16),
```

Comment:
note: `fromHeaderData` accepts `BufferLike` and `BNLike` instances and converts them, so no need to do it here.

Reply:
Oh this is a million times nicer! Thought everything needed to be pre-formatted.

Reply:
great! yeah the idea is that the constructor itself (`new BlockHeader()`) is very strict in its types to keep the initialization logic as simple as possible, but the factory helper `fromHeaderData` has flexible input types and does some minimal conversion :)

the same is true with block, tx, account, etc.